### PR TITLE
feat(coderd): add support for sending batched agent metadata

### PR DESCRIPTION
**Pull request by mafredri**
Part of #9782

### CODE REVIEW

**Comment by mafredri** [comment]
Current dependencies on/for this PR:
* main
  * **PR #10223** <a href="https://app.graphite.dev/github/pr/coder/coder/10223" target="_blank">Graphite</a>  👈
    * **PR #10224** <a href="https://app.graphite.dev/github/pr/coder/coder/10224" target="_blank">Graphite</a> 
      * **PR #10225** <a href="https://app.graphite.dev/github/pr/coder/coder/10225" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/coder/coder/10223?utm_source=stack-comment).

File: coderd/workspaceagents.go [hunk 1]
```diff
@@ -1893,42 +1904,84 @@ func (api *API) watchWorkspaceAgentMetadata(rw http.ResponseWriter, r *http.Requ
...
+			for _, datum := range md {
+				metadataMap[datum.Key] = datum
+			}
+			pendingChanges = true
+			continue
 		case <-sendTicker.C:
 			// We send an update even if there's no change every 5 seconds
 			// to ensure that the frontend always has an accurate "Result.Age".
-			if !pendingChanges.Load() && time.Since(lastSend) < time.Second*5 {
+			if !pendingChanges && time.Since(lastSend) < 5*time.Minute {
```
Comment (johnstcn) [inline coderd/workspaceagents.go:1978]: 5 minutes or 5 seconds?

File: coderd/deprecated.go [hunk 2]
```diff
@@ -70,3 +75,42 @@ func (api *API) workspaceAgentLogsDeprecated(rw http.ResponseWriter, r *http.Req
...
+// @Summary Removed: Submit workspace agent metadata
+// @ID removed-submit-workspace-agent-metadata
+// @Security CoderSessionToken
+// @Accept json
+// @Tags Agents
+// @Param request body agentsdk.PostMetadataRequestDeprecated true "Workspace agent metadata request"
+// @Param key path string true "metadata key" format(string)
+// @Success 204 "Success"
+// @Router /workspaceagents/me/metadata/{key} [post]
+// @x-apidocgen {"skip": true}
```
Comment (johnstcn) [inline coderd/deprecated.go:88]: 🤔 Is there no `@Deprecated` parameter? That's annoying

File: coderd/database/queries/workspaceagents.sql
Comment (johnstcn) [inline coderd/database/queries/workspaceagents.sql:1]: Is this new query compatible with the existing schema?

```
    key character varying(127) NOT NULL,
    value character varying(65535) DEFAULT ''::character varying NOT NULL,
    error character varying(65535) DEFAULT ''::character varying NOT NULL,
```

Comment (mafredri) [reply coderd/database/queries/workspaceagents.sql:1]: No queries or schema were changed, other than introducing batching, why would it not be compatible?

Comment (johnstcn) [reply coderd/database/queries/workspaceagents.sql:1]: Just checking, I may have misunderstood what the query was doing. I initially thought it was stuffing multiple metadata values into one, but I see now that we're just updating multiple metadata values instead.

File: coderd/workspaceagents.go [hunk 3]
```diff
@@ -1723,10 +1724,9 @@ func ellipse(v string, n int) string {
 // @Security CoderSessionToken
 // @Accept json
 // @Tags Agents
-// @Param request body agentsdk.PostMetadataRequest true "Workspace agent metadata request"
-// @Param key path string true "metadata key" format(string)
+// @Param request body []agentsdk.PostMetadataRequest true "Workspace agent metadata request"
 // @Success 204 "Success"
-// @Router /workspaceagents/me/metadata/{key} [post]
+// @Router /workspaceagents/me/metadata [post]
 // @x-apidocgen {"skip": true}
```
Comment (johnstcn) [inline coderd/workspaceagents.go:1730]: I think we need to keep the old endpoint around for legacy agents for at least one release. Older agents will continue to post to `/metadata/{key}` and get a 404.

File: coderd/deprecated.go [hunk 4]
```diff
@@ -70,3 +75,42 @@ func (api *API) workspaceAgentLogsDeprecated(rw http.ResponseWriter, r *http.Req
...
+func (api *API) workspaceAgentPostMetadataDeprecated(rw http.ResponseWriter, r *http.Request) {
```
Comment (johnstcn) [inline coderd/deprecated.go:89]: Would it be worthwhile dropping a WARN log advising operators to restart the workspace?

Comment (mafredri) [reply coderd/deprecated.go:89]: Good question. What if it's a workspace that is running something semi-critical and can't be shut down for a week? We'd potentially be spamming WARN every second :/

Comment (johnstcn) [reply coderd/deprecated.go:89]: Fair point!

**Review by johnstcn — APPROVED** [review]
This looks OK to me, but might be good to get a second approval.

**Review by mtojek — APPROVED** [review]
I haven't found anything concerning or blocking 👍

**Comment by mafredri** [comment]
### Merge activity

* **Oct 13, 9:37 AM**: @mafredri merged this pull request with [Graphite](https://app.graphite.dev/github/pr/coder/coder/10223).

**Pull request merged**